flocker/node/_deploy.py

```diff
@@ -405,13 +427,22 @@ def __init__(self, hostname, volume_service, docker_client=None,
         self.network = network
         self.volume_service = volume_service
 
-    def discover_local_state(self):
+    def discover_local_state(self, cluster_state):
         """
         List all the ``Application``\ s running on this node.
 
-        :returns: A ``Deferred`` which fires with a ``NodeState``
+        :param Deployment cluster_state: The last known state of the whole
+            cluster.
+
+        :returns: A ``Deferred`` which fires with a ``NodeApplications``
             instance.
         """
+        # XXX delete manifestations code that is unrelated to
+        # applications, return NodeApplications instance.  Application's
+        # attached volumes/manifestations can be calculated based on
+        # passed-in cluster_state. This has a potential race condition
+        # unfortunately, but probably not an issue in practice.
```

Comment:
The failure mode in this race condition is the container agent says "hm, these apps should be using manifestations as their volumes but they aren't... and there's no manifestations!" So what does it do? Presumably it stops them. If we're in situation where cluster state is missing info about the local node that means the agent was disconnected from the control service, perhaps for a while... a situation in which stopping the applications  (which will then soon be restarted) doesn't seem like it's end of the world.

Currently cluster state service never clears its memory of node state; in future it might expire after a while. Maybe convergence loop startup logic should be updated with "do a couple of iterations of discovery only until cluster state you get from control service includes local state". This'll make the race condition still there, but extremely unlikely... and combined with fact that it just means quick restart of containers we're probably fine?